task_manager/task_manager.cc

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+#include "task_manager.hh"
+#include "seastar/coroutine/parallel_for_each.hh"
+
+logging::logger tmlogger("task manager");
+
+task_manager::module_ptr task_manager::make_module(std::string name) {
+    auto m = seastar::make_shared<task_manager::module>(*this, name);
+    register_module(std::move(name), m);
+    return m;
+}
+
+task_manager::module_ptr task_manager::find_module(std::string module_name) {
+    auto it = _modules.find(module_name);
+    if (it == _modules.end()) {
+        throw std::runtime_error(format("module {} not found", module_name));
+    }
+    return it->second;
+}
+
+future<> task_manager::stop() noexcept {
+    if (!_modules.empty()) {
+        on_internal_error(tmlogger, "Tried to stop task manager while some modules were not unregistered");
+    }
+    return make_ready_future<>();
+}
+
+future<task_manager::foreign_task_ptr> task_manager::lookup_task_on_all_shards(sharded<task_manager>& tm, task_id tid) {
+    return task_manager::invoke_on_task<task_manager::foreign_task_ptr>(tm, tid, [] (task_ptr task) {
+        return make_ready_future<task_manager::foreign_task_ptr>(make_foreign(task));
+    });
+}
+
+future<> task_manager::invoke_on_task(sharded<task_manager>& tm, task_id id, std::function<future<> (task_manager::task_ptr)> func) {
```

Comment:
Nice trick for `future<>`.
Otherwise, I guess we need to specialize `invoke_on_task` below for the `future<>` case.